### PR TITLE
[model] rework DeepSeekV3

### PR DESCRIPTION
**Pull request by Luosuu**
rework DeepSeekV3 model definition

### CODE REVIEW

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 1]
```diff
@@ -0,0 +1,771 @@
+# Copied from https://github.com/huggingface/transformers/blob/41e5abac5cb49983a08ddef3e8645d6efd23c8f3/src/transformers/models/deepseek_v3/modeling_deepseek_v3.py
+# v4.56.2
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py]: can we summarize the list of changes from original hf code in comments here? (first comment not finished review yet. please add the summary to make the review easier)

Comment (Luosuu) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py]: yes, I just added a summary for a list of changes at the top of the file.

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+        with torch.autocast(device_type=device_type, enabled=False):  # Force float32
+            freqs = (inv_freq_expanded.float() @ position_ids_expanded.float()).transpose(1, 2)
+            emb = torch.cat((freqs, freqs), dim=-1)
+            cos = emb.cos() * self.attention_scaling
+            sin = emb.sin() * self.attention_scaling
+
+        return cos.to(dtype=x.dtype), sin.to(dtype=x.dtype)
+
+
+class DeepseekV3MLP(nn.Module):
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:105]: why not use the LigerSwiGLUMLP to replace the entire module like qwen3 did https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/models/transformers/qwen3/modeling_qwen3.py#L877?

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+        expert_hit = torch.greater(expert_mask.sum(dim=(-1, -2)), 0).nonzero()
+        for expert_idx in expert_hit:
+            idx, top_x = torch.where(expert_mask[expert_idx].squeeze(0))
+            current_state = hidden_states[None, top_x].reshape(-1, hidden_states.shape[-1])
+            current_hidden_states = self[expert_idx](current_state) * top_k_weights[top_x, idx, None]
+            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+        return final_hidden_states
+
+
+class DeepseekV3FusedMoe(nn.Module):
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:175]: nit: add modification comment?

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+    """
+    Fused MoE
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py]: add note that we need merged weights ckpt when using fused?

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+            requires_grad=True,
+        )
+        self.up_proj = torch.nn.Parameter(
+            torch.empty(self.num_experts, self.intermediate_size, self.hidden_dim),
+            requires_grad=True,
+        )
+        self.down_proj = torch.nn.Parameter(
+            torch.empty(self.num_experts, self.hidden_dim, self.intermediate_size),
+            requires_grad=True,
+        )
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:197]: can we assert that the config.act_fn is actually silu? (i know dpsk always uses silu just to be sure)

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 6]
```diff
@@ -0,0 +1,777 @@
...
+
+    def forward(
+        self, hidden_states: torch.Tensor, top_k_index: torch.Tensor, top_k_weights: torch.Tensor
+    ) -> torch.Tensor:
+        out = fused_moe_forward(
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:202]: did not look into detail. but why is it different from https://github.com/ByteDance-Seed/VeOmni/blob/main/veomni/models/transformers/qwen3_moe/modeling_qwen3_moe.py#L184-L207

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py [hunk 7]
```diff
@@ -0,0 +1,777 @@
...
+            module=self,
+            num_experts=self.num_experts,
+            routing_weights=top_k_weights,
+            selected_experts=top_k_weights,
```
Comment (piyifan123) [inline veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py]: still top_k_weights here?

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py (hunk 2 above)
Comment (Luosuu) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:105]: this module is reused by DeepSeekV3MoE so probably we should not replace it by Liger?

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py (hunk 5 above)
Comment (Luosuu) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:197]: should we add assert here or in the DeepseekV3MLP?

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py (hunk 6 above)
Comment (Luosuu) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py:202]: previously everything is in one MoE class but now we break it down into DeepseekV3MoE and DeepseekV3NaiveMoe/DeepseekV3FusedMoe

File: veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py (hunk 7 above)
Comment (Luosuu) [reply veomni/models/transformers/deepseek_v3/modeling_deepseek_v3.py]: typo. addressed

**Comment by Luosuu** [comment]
After discussion, we decided to work on fix current deepseekv3 modeling instead of matching it with transformers 4.57.0 since the official safetensors may do not match the latest transformers code

new branch: https://github.com/ByteDance-Seed/VeOmni/pull/new/tianle/fix-dpsk-v3
